### PR TITLE
fixes #349: Step instantiation via Step.Factory retrieved from ServiceLoader

### LINKED ISSUE #349: Steps shall be instantiated via Step.Factory methods
### Expected Behavior
A Step is instantiated via a Step.Factory.

### Current Behavior
Steps are instantiated by calling the default constructor.

### PR DESCRIPTION
**Pull request by mklaehn**

### CODE REVIEW

File: controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java [hunk 1]
```diff
@@ -24,18 +24,57 @@
...
+
+    default void prepareStep(final MachineContext context) {
+    }
+
+    void doStep(final MachineContext context);
+
+    default void leaveStep(final MachineContext context) {
+    }
+
+    Duration preferredStepDuration(final MachineContext context);
```
Comment (svenreimers) [inline controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java]: Should preferredStepDuration have a default implementation, e.g Duration.ZERO?

File: controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java [hunk 2]
```diff
@@ -24,18 +24,57 @@
...
+        return true;
+    }
+
+    /**
+     * A Factory creating a {@link DataProvider}.
+     */
+    interface Factory {
+
+        /**
+         * Returns the class of the Provider this factory will create via
```
Comment (svenreimers) [inline controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java]: Provider should be Step in the JavaDoc

File: controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java [hunk 3]
```diff
@@ -24,18 +24,57 @@
...
+         * {@link #create(org.tweetwallfx.tweet.api.TweetStream)}.
+         *
+         * @return the class of the Provider this factory will create
```
Comment (svenreimers) [inline controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java]: Provider should be Step

File: controls/src/main/java/org/tweetwallfx/controls/stepengine/StepIterator.java [hunk 4]
```diff
@@ -101,19 +104,48 @@ public Step next() {
...
-        public Builder addStep(final String classname) {
-            try {
-                Object newInstance = Thread.currentThread().getContextClassLoader().loadClass(classname).getDeclaredConstructor().newInstance();
-                if (newInstance instanceof Step) {
-                    steps.add((Step) newInstance);
+        public Builder addStep(final String stepClassName) {
+            final Step.Factory factory = FACTORIES.get(stepClassName);
+
+            if (null == factory) {
+                LOGGER.error("No Factory exist that can create Step '{}'", stepClassName);
```
Comment (svenreimers) [inline controls/src/main/java/org/tweetwallfx/controls/stepengine/StepIterator.java]: Typo: exist -> exists

File: controls/src/main/java/org/tweetwallfx/controls/stepengine/StepIterator.java [hunk 5]
```diff
@@ -101,19 +104,48 @@ public Step next() {
...
+            } else {
+                final Step step = factory.create();
+
+                if (null == step) {
```
Comment (svenreimers) [inline controls/src/main/java/org/tweetwallfx/controls/stepengine/StepIterator.java]: Do we want to allow null as return of factory.create? Feels weird.

Comment (mklaehn) [reply controls/src/main/java/org/tweetwallfx/controls/stepengine/StepIterator.java]: I'm good with throwing an exception...

File: controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java (hunk 1 above)
Comment (mklaehn) [reply controls/src/main/java/org/tweetwallfx/controls/stepengine/Step.java]: okay for me. @jmichelberger what do you think?

**Pull request merged**